### PR TITLE
Adapt links to cantonal geoportals

### PR DESCRIPTION
**Pull request by nadine-piveteau**
See[ issue #4020](https://github.com/geoadmin/mf-geoadmin3/issues/4020#issuecomment-341657468)

### CODE REVIEW

File: chsdi/templates/htmlpopup/cadastralwebmap.mako [hunk 1]
```diff
@@ -10,7 +10,7 @@
   from chsdi.models.vector import get_scale
   from chsdi.lib.validation.identify import IdentifyServiceValidation
   request = context.get('request')
-  grid = getTileGrid(21781)()
+  grid = getTileGrid(2056)()
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap.mako]: pass in params.srid

File: chsdi/templates/htmlpopup/cadastralwebmap.mako [hunk 2]
```diff
@@ -22,12 +22,14 @@
...
-  c['bboxlv95'] = [c['bbox'][0] + 2000000, c['bbox'][1] + 1000000,
-                   c['bbox'][2] + 2000000, c['bbox'][3] + 1000000]
+  c['bboxlv03'] = [c['bbox'][0] - 2000000, c['bbox'][1] - 1000000,
+                   c['bbox'][2] - 2000000, c['bbox'][3] - 1000000]
   c['scale']  = get_scale(params.imageDisplay, params.mapExtent)
   defaultCoord = [(c['bbox'][0] + c['bbox'][2]) / 2,
                   (c['bbox'][1] + c['bbox'][3]) / 2]
   clickCoord = request.params.get('coord').split(',') if request.params.get('coord') else defaultCoord
+  c['clickCoordLv03'] = [clickCoord[0][1:],
+                         clickCoord[1][1:]]
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap.mako]: why do you use [1:] ? coords in this case is also a list of 2 coordinates which srid is determined by `params.srid`
so make sure you clearly have both `clickCoordLv03` and `clickCoordLv95` in the Mako context because `clickCoord` could be either LV95 or LV03.

File: chsdi/templates/htmlpopup/cadastralwebmap.mako [hunk 3]
```diff
@@ -22,12 +22,14 @@
   params = CadastralWebMapParams(request)
 
   c['bbox'] = params.mapExtent.bounds
-  c['bboxlv95'] = [c['bbox'][0] + 2000000, c['bbox'][1] + 1000000,
-                   c['bbox'][2] + 2000000, c['bbox'][3] + 1000000]
+  c['bboxlv03'] = [c['bbox'][0] - 2000000, c['bbox'][1] - 1000000,
+                   c['bbox'][2] - 2000000, c['bbox'][3] - 1000000]
```
[the remark this answers is not in the document]
Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: you could simply replace bbox with `bboxlv95` and `bboxlv03`, thus passing explicitly the correct values in the mako.

[the remark this answers is not in the document]
Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: FYI: this method `shift_to(coords, srid)` will transform both point and bbox coordinates to the correct value.

https://github.com/geoadmin/mf-chsdi3/blob/master/chsdi/lib/helpers.py#L316

File: chsdi/templates/htmlpopup/cadastralwebmap.mako (hunk 2 above)
Comment (nadine-piveteau) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: clickCoord returns a tuple of strings which represents the coordinates in LV95. As I needed the coordinates in LV03 I removed the 2's (east) and the 1's (north) by shortening the string by its first character. I agree, it's not a very nice hack :D 
Anyway, I will in a next step add it to the context variable such that it is not necessary anymore to apply that .

**Comment by nadine-piveteau** [comment]
Thx for the feedback! I added the requested modifications. Because of these changes I also had to modify the code for OpenData-AV. Thus, I would be grateful if you could test both, CadastralWebMap and OpenData-AV. Thx in advance!
(There are two tests concerning Sphinx and the search service that failed. I think it has nothing to do with my modifications, but let me know, if I should fix that!)
 
[Testlink](https://mf-geoadmin3.dev.bgdi.ch/ltpin/src/?api_url=%2F%2Fmf-chsdi3.int.bgdi.ch%2Fltpin_cadastralwebmap&lang=de&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=ch.kantone.cadastralwebmap-farbe,ch.swisstopo-vd.amtliche-vermessung&E=2692010.84&N=1195546.62&zoom=2)

@AFoletti : I've also updated the link to SG-Geoportal.

**Comment by nadine-piveteau** [comment]
I just realized that some PDF-Download links do not work anymore given that some cantons expect LV03 coordinates. Working on that...

**Comment by loicgasser** [comment]
ok thanks, it's a bit of a juggle between lv95 and lv03 for cantons as well.

**Comment by loicgasser** [comment]
sphinx is because of some on going devs on search server. (missing quadindex for locations_preview)
so unrelated

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 4]
```diff
@@ -12,35 +12,49 @@
...
   class CadastralWebMapParams(IdentifyServiceValidation):
       def __init__(self, request):
+          self.srid = request.params.get('sr', '21781')
+          grid = getTileGrid(self.srid)()
+          defaultExtent = ','.join(map(str, grid.extent))
+          defaultImageDisplay = '400,600,96'
           self.mapExtent = request.params.get('mapExtent', defaultExtent)
           self.imageDisplay = request.params.get('imageDisplay', defaultImageDisplay)
+          self.coord = request.params.get('coord')
+  toto
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: toto?

File: chsdi/templates/htmlpopup/cadastralwebmap.mako [hunk 5]
```diff
@@ -9,25 +9,30 @@
...
           self.mapExtent = request.params.get('mapExtent', defaultExtent)
           self.imageDisplay = request.params.get('imageDisplay', defaultImageDisplay)
-  params = CadastralWebMapParams(request)
+          self.coord = request.params.get('coord')
 
-  c['bbox'] = params.mapExtent.bounds
-  c['bboxlv95'] = [c['bbox'][0] + 2000000, c['bbox'][1] + 1000000,
-                   c['bbox'][2] + 2000000, c['bbox'][3] + 1000000]
+  params = CadastralWebMapParams(request)
+  c['bboxlv95'] =  list(params.mapExtent.bounds)
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap.mako]: if sr parameter == 2056 then this is incorrect...
why did you remove the if we added yesterday?

Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: 21781 I mean

Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: and you need the fix in geoadmin otherwise it won't work.

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 6]
```diff
@@ -12,35 +12,49 @@
...
   params = CadastralWebMapParams(request)
-  c['bbox'] = params.mapExtent.bounds
-  c['bboxlv95'] = [2000000 + c['bbox'][0], 1000000 + c['bbox'][1], 2000000 + c['bbox'][2], 1000000 + c['bbox'][3]]
+  c['bboxlv95'] = list(params.mapExtent.bounds)
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: same here:
```python
if params.srid == 2056:
    c['bboxlv95'] = list(params.mapExtent.bounds)
    c['bboxlv03'] = shift_to(c['bboxlv95'], 21781)
else:
    c['bboxlv03'] = list(params.mapExtent.bounds)
    c['bboxlv95'] = shift_to(c['bboxlv03'], 2056)
```

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 7]
```diff
@@ -12,35 +12,49 @@
...
   c['scale']  = get_scale(params.imageDisplay, params.mapExtent)
   topic = request.matchdict.get('map')
   baseUrl = request.registry.settings['api_url']
-  coord = request.params.get('coord').split(',') if request.params.get('coord') else defaultCoord
-  lat = coord[0]
-  lon = coord[1]
+
+  defaultCoord = [(c['bboxlv95'][0]+c['bboxlv95'][2])/2,
+                  (c['bboxlv95'][1]+c['bboxlv95'][3])/2]
+  c['clickCoordLv95'] = [float(a) for a in params.coord.split(',')] if params.coord else defaultCoord
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: here you need to check srid as well.

Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: ah no that's ok

File: chsdi/templates/htmlpopup/cadastralwebmap.mako (hunk 5 above)
Comment (nadine-piveteau) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: Yes I removed it because I noticed that the params coord and mapExtent are delivered per default in Lv95 also when sr=21781.

Comment (loicgasser) [reply chsdi/templates/htmlpopup/cadastralwebmap.mako]: no, incorrect default is lv03 still, you have to pass in sr parameter....
If you tested with prod it's normal it didn't work.

File: chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako [hunk 8]
```diff
@@ -12,35 +12,49 @@
...
+  c['clickCoordLv03'] = shift_to(c['clickCoordLv95'], 21781)
+
+  if params.srid == '2056':
+    lat = c['clickCoordLv95'][0]
+    lon = c['clickCoordLv95'][1]
+  else:
+    lat = c['clickCoordLv03'][0]
+    lon = c['clickCoordLv03'][1]
```
Comment (loicgasser) [inline chsdi/templates/htmlpopup/cadastralwebmap_opendata.mako]: this should always be lv03...

**Comment by nadine-piveteau** [comment]
yes I redeployed the branch to int.

**Comment by loicgasser** [comment]
thanks!
